src/NHSD.GPIT.BuyingCatalogue.Services/Competitions/CompetitionsService.cs

```
@@ -655,37 +805,90 @@ public async Task<bool> Exists(string internalOrgId, string competitionName) =>
+        {
+            throw new ArgumentException(@"competitionSublocations is null or empty", nameof(competitionSublocations));
+        }
 
-        dbContext.CompetitionRecipients.RemoveRange(staleRecipients);
-        dbContext.CompetitionRecipients.AddRange(newRecipients.Select(x => new CompetitionRecipient(competitionId, x)));
+        Competition competition = await dbContext.Competitions
+            .Where(x => x.Organisation.InternalIdentifier == internalOrgId && x.Id == competitionId)
+            .Include(x => x.Organisation)
+            .Include(x => x.CompetitionSublocations)
+            .ThenInclude(y => y.SublocationRecipients)
+            .FirstAsync();
 
+        if (competition.Completed.HasValue)
+        {
+            throw new InvalidOperationException(
+                "Cannot set sublocations / recipients on a completed competition.");
+        }
+
+        IReadOnlyList<OdsOrganisation> validSublocations =
+            await odsService.GetSublocationsByParentOdsCode(competition.Organisation.ExternalIdentifier);
+
+        var validateSublocations = competitionSublocations
+            .All(x => validSublocations.Select(y => y.OdsCode).Contains(x.SublocationOdsCode));
+
+        if (!validateSublocations)
+        {
+            throw new InvalidOperationException("Provided sublocations not valid for this organisation.");
+        }
+
+        var validateRecipients = true;
+
+        foreach (CompetitionSublocation sublocation in competitionSublocations)
+        {
+            IReadOnlyList<ServiceRecipient> validServiceRecipients =
+                await odsService.GetServiceRecipientsBySublocation(sublocation.SublocationOdsCode);
+
+            var instanceCheck = sublocation.SublocationRecipients.All(
+                x => validServiceRecipients.Select(y => y.OrgId).Contains(x.RecipientOdsCode));
+
+            if (!instanceCheck)
+            {
+                validateRecipients = false;
```

In the spirit of failing fast, the exception could be raised here at the first sign of a sublocation recipient being invalid since the outcome would be the same. Then the additional boolean isn't required

---

Moved throw to within iteration block